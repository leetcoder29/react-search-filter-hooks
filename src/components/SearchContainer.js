import React, { createContext, useReducer } from 'react';

import data from '../db/data';
import SearchInput from './SearchInput';
import SearchResults from './SearchResults';
import SearchFilter from './SearchFilter';
/**
 * 1. Create a SearchContainer => context provider for application state
 * ✅ Create context and make top level div the provider
 * ☑️ Create reducer hook for state changes
 * - SUBMIT_SEARCH
 * - SET_FILTER
 * ✅ Create action handlers for search submit
 */

// React Context
export const SearchContext = createContext(null);

// React useReducer Hook
const searchReducer = (state, action) => {
  switch (action.type) {
    case 'SUBMIT_SEARCH':
      return {
        ...state,
        searchResults: action.payload,
      };

    case 'SET_FILTER':
      return {
        ...state,
        filterKey: action.payload,
      };

    default:
      return state;
  }
};

const initialState = {
  searchTerm: '',
  searchResults: [],
  filterKey: 'NONE',
};

export const getSearchResults = (data, searchTerm) => {
  return data.filter(item => {
    return item['first_name'].toLowerCase() === searchTerm;
  });
};

const SearchContainer = () => {
  const [searchState, dispatchSearch] = useReducer(searchReducer, initialState);

  const handleSearchSubmit = searchTerm => {
    // Filter data by first name according to searchTerm
    const results = getSearchResults(data, searchTerm);

    return dispatchSearch({
      type: 'SUBMIT_SEARCH',
      payload: results,
    });
  };

  const handleSetFilter = filterKey => {
    return dispatchSearch({
      type: 'SET_FILTER',
      payload: filterKey,
    });
  };

  return (
    <SearchContext.Provider
      value={{
        ...searchState,
        dispatchSearch,
        handleSearchSubmit,
        handleSetFilter,
      }}
    >
      <h1>Simple Search and Filter</h1>
      <SearchInput />
      <SearchFilter />
      <SearchResults />
    </SearchContext.Provider>
  );
};

export default SearchContainer;
