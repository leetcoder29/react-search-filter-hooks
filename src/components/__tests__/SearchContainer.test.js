import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJSON from 'enzyme-to-json';

import data from '../../db/data';
import SearchContainer, { getSearchResults } from '../SearchContainer';

configure({ adapter: new Adapter() });

describe('Search functionality', () => {
  it('returns filtered array of search results from data with a valid search term', () => {
    const searchTerm = 'hi';
    const searchResults = getSearchResults(data, searchTerm);
    expect(searchResults[0]['first_name'].toLowerCase()).toEqual(searchTerm);
  });

  it('returns an empty array of search results from data with an invalid search term', () => {
    const searchTerm = 'bruce';
    const searchResults = getSearchResults(data, searchTerm);
    expect(searchResults).toHaveLength(0);
  });
});

describe('<SearchContainer /', () => {
  it('renders correctly', () => {
    shallow(<SearchContainer />);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<SearchContainer />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
