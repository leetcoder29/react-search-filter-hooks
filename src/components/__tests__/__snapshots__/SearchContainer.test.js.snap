// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchContainer / matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "dispatchSearch": [Function],
      "filterKey": "NONE",
      "handleSearchSubmit": [Function],
      "handleSetFilter": [Function],
      "searchResults": Array [],
      "searchTerm": "",
    }
  }
>
  <h1>
    Simple Search and Filter
  </h1>
  <SearchInput />
  <SearchFilter />
  <SearchResults />
</ContextProvider>
`;
